Constraint analysis scripts for initial sizing

stankarpuktubs opened the pull request:
# Pull Request Checklist
 * Merge in the latest Develop branch changes to your branch
 * Remove .pyc files from your repository:
   * Linux/Mac
     * find . -name *.pyc -delete
   * Windows
     * del /S *.pyc
 * Run automatic regression and makes sure everything is passing
 * Did you make a new regression test that covers your new code?
 * Did you update your docstrings?
 * Did you update your headers to include your name and date?
 * Do a final compare with the Develop branch to be sure what you're changing

stankarpuk93 committed:
Constraint analysis scripts for initial sizing

stankarpuk93 committed:
changed the regression directory

stankarpuk93 committed:
Updated the constraint diagram

A deep methods update and a regression script

stankarpuk93 committed:
regression update

stankarpuktubs committed:
Updated constraint analysis regression

Updated the propeller aircraft constraint analysis regression

stankarpuktubs committed:
Update automatic_regression.py

stankarpuktubs committed:
Merge branch 'develop' into develop

stankarpuktubs committed:
Removed ADRpy

stankarpuktubs committed:
Merge branch 'develop' of https://github.com/stankarpuktubs/SUAVE into develop

stankarpuktubs committed:
Updated regression

stankarpuktubs committed:
Corrected the regression

stankarpuktubs committed:
Update constraint_analysis.py

stankarpuktubs committed:
Checked the code structure and extended regression cases

stankarpuktubs committed:
Update Constraint_Analysis.py

stankarpuktubs committed:
Update constraint_analysis.py

planes commented on trunk/SUAVE/Analyses/Constraint_Analysis/Constraint_Analysis.py:
```diff
@@ -0,0 +1,1386 @@
+            T_W[i] = q*cd0/W_S[i]+k*W_S[i]/q
+
+        # Convert thrust to power (for propeller-driven engines) and normalize the results wrt the Sea-level
+        if eng_type != 'turbofan' and eng_type != 'turbojet':
```
Much of this code is copy and paste from other functions. Find ways to condense code.

This will help in several ways. It will reduce the size of this file, but most importantly it serves to reduce bugs.

planes commented on trunk/SUAVE/Analyses/Constraint_Analysis/Constraint_Analysis.py:
```diff
@@ -0,0 +1,1386 @@
+        # Unpack inputs
+        eng_type  = self.engine.type
+        cd_min    = self.aerodynamics.cd_min_clean 
+        altitude  = self.climb.altitude   
```
is this a scalar value or a vector?

planes commented on trunk/SUAVE/Analyses/Constraint_Analysis/Constraint_Analysis.py:
```diff
@@ -0,0 +1,1386 @@
+        P         = K*cdmin
+        Q         = 1/(kf*u)
+
+        #if mach > 0.3:
```
remove commented out code

planes commented on trunk/SUAVE/Analyses/Constraint_Analysis/Constraint_Analysis.py:
```diff
@@ -0,0 +1,1386 @@
+        return e
+
+
+    def estimate_max_lift(self,highlift_type):
```
I think this could be useful for many other calculations. Can you make this more general and move to methods?

stankarpuktubs replied:
I see that SUAVE already has the maximum lift estimation function. Is it better to have this function as a stand-alone or to integrate it into the SUAVE default maximum lift function?

planes replied:
If you're able to integrate it, that would be great. You can make a setting that will switch between types.

planes commented on trunk/SUAVE/Analyses/Constraint_Analysis/Constraint_Analysis.py:
```diff
@@ -0,0 +1,1386 @@
+            raise ValueError(highliftmsg)
+
+
+    def compressibility_drag(self,mach,cl):
```
call the existing scripts since this is much the same

stankarpuktubs replied:
Here, I am struggling with the fundamental difference between functions of SUAVE and requirements for the constraint diagram. All default functions require definitions of settings, states, etc, since thez are created for the mission segment analysis. Here, I just need 2-3 inputs for the execution with minimum input, so the constraint diagram can be initialized and executed as simple as possible. For the latest update, I still have not managed to integrate these functions into the default SUAVE

planes replied:
I understand you're trying to do a simple analysis here. I'm not trying to change that. I bet you think this is frustrating because you coded this with a certain intention of this being simple.

For this function, you can easily replace it with the [original](https://github.com/suavecode/SUAVE/blob/develop/trunk/SUAVE/Methods/Aerodynamics/Common/Fidelity_Zero/Drag/compressibility_drag_wing.py). Settings are not used, hence they become None. Geometry only requires wing that has a thickness to chord and sweep. State is just a dictionary that saves the outputs and provides a CL. It's not tricky to use the existing script.

We've found that having reproduced code eventually leads to bugs and just bloat. It's more confusing and it all becomes a mess eventually.

planes reviewed (commented):
The two big overall comments:
Split things out into methods and make consistent with SUAVE vehicle data structures.

Overall good work. I think this is a good contribution.

stankarpuktubs committed:
Modified the code structure. Removed unnecessary stuff.

planes commented:
If you haven't seen already, there's an issue with versions of a Python package unrelated to SUAVE. This is giving the red X. When the issue is fixed, I'll rebuild and we can finish off the pull request.

stankarpuktubs commented on trunk/SUAVE/Analyses/Constraint_Analysis/Constraint_Analysis.py:
```diff
@@ -0,0 +1,1386 @@
+        g           = atmosphere.planet.sea_level_gravity
+
+        T_W = np.zeros(len(W_S))
+        if eng_type != 'turbofan' and eng_type != 'turbojet':
```
I see. The main goal of a constraint diagram here is to enable the analzsis with minimum anount of initial definitions, so one can essentially run it without a vehicle and build a vehicle later, for example. I also have stuff like air-cooled and liquid-cooled electric motors. Shall I still change this structure?

planes replied:
I understand the constraint diagram should be as simple as possible. The change I want makes this all simpler actually.

All the user needs to do use this as I want is to append a network. The network can be totally empty. The user doesn't need to know the exact propulsion names that this script chose. 

Then as they move through the design process they can start adding more detail to that network and running missions. Similarly the wing definition can be improved (adding segments and airfoils etc...). 

The cool thing is, as they provide further data, they can always go back to the constraint diagram code. In practice for a designer making airplanes their process can be iterative and nonlinear. This prevents them from having to make two totally disconnected SUAVE models to use different analyses.

planes commented on regression/scripts/constraint_analysis/constraint_analysis.py:
```diff
@@ -44,36 +49,28 @@ def main():
-    ca.geometry.high_lift_type_clean   = None
-    ca.geometry.high_lift_type_takeoff = 'double-slotted Fowler'
-    ca.geometry.high_lift_type_landing = 'double-slotted Fowler'
+    ca.geometry.aspect_ratio                 = vehicle.wings['main_wing'].aspect_ratio
```
I was hoping it would look something like this:

ca.geometry = B737_vehicle()

That way a user doesn't need to do all of this manually. 

I think a really neat workflow for a new user who would be interested in your contributions here could be to draw a vehicle in OpenVSP and very quickly produce the constraint diagram. That would enable this.

stankarpuktubs replied:
I am actually working on the neat integration now, so the constraint diagram script will accept the entire vehicle with some pre-defined constraint diagram settings. I can also do it the way you suggested. Hope, I will commit the deep update by the end of next week

planes replied:
Great! Do what you think makes most sense here.

I'm looking forward to this. I think your work here fills an important need in SUAVE.

planes commented on trunk/SUAVE/Analyses/Constraint_Analysis/Constraint_Analysis.py:
```diff
@@ -12,6 +12,14 @@
+from SUAVE.Methods.Constraint_Analysis.compute_cruise_constraint    import compute_cruise_constraint
+from SUAVE.Methods.Constraint_Analysis.compute_ceiling_constraint   import compute_ceiling_constraint
+from SUAVE.Methods.Constraint_Analysis.compute_landing_constraint   import compute_landing_constraint
+from SUAVE.Input_Output.Results.plot_constraint_diagram                     import plot_constraint_diagram
```
Thanks!

planes commented on trunk/SUAVE/Input_Output/Results/plot_constraint_diagram.py:
```diff
@@ -0,0 +1,105 @@
+    f.close()    
+
+
+# ----------------------------------------------------------------------
```
Delete this lower bit. This is our fault for leaving this in our style.

planes replied:
This

planes reviewed (commented):
Just adding a few new comments.

stankarpuktubs committed:
Modification of the constraint analysis

1. Moved the geometry and energy network definition into the Vehicle (where was possible)
2. Defined the constraint analysis variable set within the vehicle
3. Allocated analyses scripts within the Aero folder
4.  Allocated the post-processing code into the Input/Output folder
5. Fixed many parts to make the structure work

stankarpuktubs committed:
Corrected the analysis initialization script

stankarpuktubs committed:
Corrected Methods init

stankarpuktubs committed:
Added cases to improve coverage

stankarpuktubs committed:
Update constraint_analysis.py

stankarpuktubs committed:
Merge branch 'develop' into develop

stankarpuktubs commented:
Emilio,

I think, I have changed almost all parts where needed and embedded the constraint analyses definitions into the vehicle, so, as you described, the designer can start from minimum data and then build the vehicle step-by-step.

planes commented on trunk/SUAVE/Input_Output/Results/plot_constraint_diagram.py:
```diff
@@ -0,0 +1,103 @@
+## @ingroup Input_Output-Results
```
this should be moved to this folder: https://github.com/suavecode/SUAVE/tree/develop/trunk/SUAVE/Plots/Performance

planes commented on trunk/SUAVE/Methods/Aerodynamics/Common/Fidelity_Zero/Helper_Functions/oswald_efficiency.py:
```diff
@@ -0,0 +1,62 @@
+    """  
+
+    # Unpack inputs
+    taper = vehicle.wings['main_wing'].taper
```
I would prefer if this parsed for the main wing rather than look for a name. The code is simple:

from SUAVE.Components.Wings.Main_Wing

main_wing = None
for wing in wings:
	if isinstance(wing,main_wing):
		main_wing = wing

planes commented on trunk/SUAVE/Methods/Constraint_Analysis/compressibility_drag_constraint.py:
```diff
@@ -0,0 +1,78 @@
+## @ingroup Methods-Constraint_Analysis
+# Oswald_efficiency.py
```
This header is wrong

planes commented on trunk/SUAVE/Methods/Constraint_Analysis/compressibility_drag_constraint.py:
```diff
@@ -0,0 +1,78 @@
+import numpy as np
+
+# ------------------------------------------------------------------------------------
+#  Compute maximum lift coefficient for the constraint analysis
```
This title is wrong

planes commented on trunk/SUAVE/Methods/Constraint_Analysis/compressibility_drag_constraint.py:
```diff
@@ -0,0 +1,78 @@
+# ------------------------------------------------------------------------------------
+
+## @ingroup Methods-Constraint_Analysis
+def compressibility_drag_constraint(mach,cl,geometry):
```
This is a carbon copy of an existing script: https://github.com/suavecode/SUAVE/blob/develop/trunk/SUAVE/Methods/Aerodynamics/Common/Fidelity_Zero/Drag/compressibility_drag_wing.py

Remove this script and reference the existing one

planes commented on trunk/SUAVE/Methods/Constraint_Analysis/compute_ceiling_constraint.py:
```diff
@@ -0,0 +1,121 @@
+        T_W[i] = 0.508/(np.sqrt(2/rho*W_S[i]*np.sqrt(k/(3*cd0)))) + 4*np.sqrt(k*cd0/3)
+
+    # Convert thrust to power (for propeller-driven engines) and normalize the results wrt the Sea-level       
+    if eng_type != ('turbofan' or 'Turbofan') and eng_type != ('turbojet' or 'Turbojet'):
```
this should check for an instance of a network

planes commented on trunk/SUAVE/Methods/Constraint_Analysis/compute_climb_constraint.py:
```diff
@@ -0,0 +1,147 @@
+    Vx        = vehicle.constraints.analyses.climb.airspeed   
+    Vy        = vehicle.constraints.analyses.climb.climb_rate                   
+    delta_ISA = vehicle.constraints.analyses.climb.delta_ISA
+    AR        = vehicle.wings['main_wing'].aspect_ratio 
```
parse for the main_wing rather than relying on a name

planes commented on trunk/SUAVE/Methods/Constraint_Analysis/compute_climb_constraint.py:
```diff
@@ -0,0 +1,147 @@
+            T_W[i] = Vy/Vx + q/W_S[i]*cd0 + k/q*W_S[i]
+            P_W[i] = T_W[i]*Vx/etap
+
+            if eng_type == ('turboprop' or 'Turboprop'):
```
check for instances

planes commented on trunk/SUAVE/Methods/Constraint_Analysis/compute_cruise_constraint.py:
```diff
@@ -0,0 +1,129 @@
+    eng_type  = vehicle.constraints.engine.type
+    cd_min    = vehicle.constraints.aerodynamics.cd_min_clean 
+    W_S       = vehicle.constraints.wing_loading
+    AR        = vehicle.wings['main_wing'].aspect_ratio
```
parse

planes commented on trunk/SUAVE/Methods/Constraint_Analysis/compute_turn_constraint.py:
```diff
@@ -0,0 +1,129 @@
+        if etap == 0:
+            raise ValueError('Warning: Set the propeller efficiency during turn')
+
+    for i in range(len(W_S)):
```
2 things I see.

First, this section of code is used repeatedly in several scripts. This indicates this could be functionalized.

Second, I believe you could vectorize this for loop.

planes commented on trunk/SUAVE/Vehicle.py:
```diff
@@ -378,3 +380,137 @@ def get_children(self):
+        self.plot_tag           = False
+
+
+        # Defines default constraint analyses
```
vehicle.analyses breaks a fundamental assumption of SUAVE having data/analysis orthogonality

stankarpuktubs replied:
Unfortunately, I did not get the issue here. Could you explain it a bit more?

planes commented on trunk/SUAVE/Vehicle.py:
```diff
@@ -378,3 +380,137 @@ def get_children(self):
+        self.propeller.OEI_climb_efficiency = 0.0
+
+        # Define aerodynamics
+        self.aerodynamics = Data()
```
This is basically an analysis

stankarpuktubs committed:
Deep update

Updated the data constraint analysis structure
Fixed all errors, relocated useful files
Added a dummy turboprop energy network (similar to ICE, but with a different power lapse function)

stankarpuktubs committed:
Dates update

stankarpuktubs commented:
It seems like the appveyor issue, right?

stankarpuktubs committed:
Changed the compressibility drag calculation reference

stankarpuktubs committed:
Removed cache import

stankarpuktubs committed:
Extra formating

stankarpuktubs commented:
Have updated the analyses, used existing scripts as much as possible, and corrected many small things in the code

stankarpuktubs commented:
> Have updated the analyses, used existing scripts as much as possible, and corrected many small things in the code

stankarpuktubs commented on trunk/SUAVE/Components/Energy/Networks/Turboprop_Constant_Speed.py:
```diff
@@ -0,0 +1,158 @@
+#  Network
+# ----------------------------------------------------------------------
+## @ingroup Components-Energy-Networks
+class Turboprop_Constant_Speed(Network):
```
I see. Indeed this script is identical to the ICE. The only difference is inside the  turboprop converter (https://github.com/stankarpuktubs/SUAVE/blob/develop/trunk/SUAVE/Components/Energy/Converters/Turboprop.py) where I am using the power lapse formulation for a turboprop. Shall I then modify the ICE network to work with the turboprop converter to avoid unnecessary copies of the code?

stankarpuktubs committed:
Corrected grammar and minor errors

stankarpuktubs committed:
Merge branch 'develop' into develop

stankarpuktubs committed:
Updates of the Internal Combustion Energy Network

Merged a sample turboprop and an ICE into one energy network.
modified the constraint analysis script to include changes of the energy network

stankarpuktubs committed:
Merge branch 'develop' of https://github.com/stankarpuktubs/SUAVE into develop
